lightning/backend/local.go

```diff
@@ -1510,3 +1498,181 @@ func (s *sizeProperties) iter(f func(p *rangeProperty) bool) {
 		return f(prop)
 	})
 }
+
+type LocalWriter struct {
+	db                *pebble.DB
+	writeErr          common.OnceError
+	local             *LocalFile
+	lastKey           []byte
+	produceWg         sync.WaitGroup
+	consumeWg         sync.WaitGroup
+	kvsChan           chan []common.KvPair
+	sstDir            string
+	memtableSizeLimit int64
+}
+
+func (w *LocalWriter) isSorted(kvs []common.KvPair) bool {
+	if len(kvs) <= 1 {
+		return false
+	}
+	for _, pair := range kvs {
+		if len(w.lastKey) > 0 && bytes.Compare(w.lastKey, pair.Key) >= 0 {
+			return false
+		}
+		w.lastKey = pair.Key
+	}
+	l := len(w.lastKey)
+	lastKey := make([]byte, l)
+	copy(lastKey, w.lastKey)
+	w.lastKey = lastKey
+	return true
+}
+
+func (w *LocalWriter) AppendRows(ctx context.Context, tableName string, columnNames []string, ts uint64, rows Rows) error {
+	kvs := rows.(kvPairs)
+	if len(kvs) == 0 {
+		return nil
+	}
+	w.kvsChan <- kvs
+	w.local.Ts = ts
+	return nil
+}
+
+func (w *LocalWriter) Close() error {
+	w.WaitConsume()
+	return w.writeErr.Get()
+}
+
+func (w *LocalWriter) Done() {
+	w.produceWg.Done()
+}
+
+func (w *LocalWriter) WaitConsume() {
+	w.produceWg.Wait()
+	close(w.kvsChan)
+	w.consumeWg.Wait()
+}
+
+func (w *LocalWriter) AddProducer() {
+	w.produceWg.Add(1)
+}
+
+func (w *LocalWriter) writeRowsLoop() {
+	batchSize := int64(0)
+	totalSize := int64(0)
+	var writer *sstable.Writer = nil
+	//var wb *pebble.Batch = nil
+	var wb []common.KvPair
+	var filePath string
+	defer w.consumeWg.Done()
+	var err error
+	for kvs := range w.kvsChan {
+		atomic.AddInt64(&w.local.Length, int64(len(kvs)))
+		if len(wb) == 0 && w.isSorted(kvs) {
+			if writer == nil {
+				writer, filePath, err = w.createWriter()
+				if err != nil {
+					w.writeErr.Set(err)
+					return
+				}
+			}
+			sz, err := writeKVs(writer, kvs)
+			if err != nil {
+				w.writeErr.Set(err)
+				return
+			}
+			totalSize += sz
+		} else {
+			for _, pair := range kvs {
+				batchSize += int64(len(pair.Key) + len(pair.Val))
+			}
+			if len(wb) == 0 {
+				wb = kvs
+			} else {
+				wb = append(wb, kvs...)
+			}
+
+			if batchSize > w.memtableSizeLimit {
+				if err := w.flushKVs(wb); err != nil {
+					w.writeErr.Set(err)
+					return
+				}
```

Comment:
Since the wb have been flushed to disk. Does the wb need to be reset? Something like `wb = wb[:0]`.

Reply:
Yes... It is a bug.